packages/core/src/StateNode.ts

```diff
@@ -1242,9 +1291,24 @@ class StateNode<
       ? currentState.configuration
       : [];
 
+    const currentInput: NonNullable<TExtra['input']> =
+      currentState?.input || this.options.input!;
+
+    let updatedInput: NonNullable<TExtra['input']> | undefined = undefined;
+
+    if (_event?.name === actionTypes.input) {
+      updatedInput = Object.assign(currentInput, (_event.data as any).input);
```

Comment:
```suggestion
      updatedInput = Object.assign({}, currentInput, (_event.data as any).input);
```

We want to keep track of previous input in `state.history` instead of mutating it

Reply:
Need to add a test to that effect before committing this

Reply:
Nice, added a test which failed without this change and passed with it. Good eyes!